### PR TITLE
Task/aa 2047 add ask screen composable

### PR DESCRIPTION
**Pull request by kelianClerc**
Add buttons composable 
Capture d’écran 2022-09-29 à 09 54 36

Add ask screen composable
Capture d’écran 2022-09-29 à 09 54 29

### CODE REVIEW

File: library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ */
+@Immutable
+private class GrapesButtonColors(
+    private val backgroundColor: Color,
+    private val contentColor: Color,
+    private val disabledBackgroundColor: Color,
+    private val disabledContentColor: Color
+) : ButtonColors {
+    @Composable
+    override fun backgroundColor(enabled: Boolean): State<Color> {
```
Comment (jpbarbaud) [inline library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:23]: When there is only 1 line, I used to inline the function like this 
 override fun backgroundColor(enabled: Boolean): State<Color>  = rememberUpdatedState(if (enabled) backgroundColor else disabledBackgroundColor)

File: library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        return rememberUpdatedState(if (enabled) backgroundColor else disabledBackgroundColor)
+    }
+
+    @Composable
+    override fun contentColor(enabled: Boolean): State<Color> {
```
Comment (jpbarbaud) [inline library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:28]: same

File: library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        return rememberUpdatedState(if (enabled) contentColor else disabledContentColor)
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as GrapesButtonColors
+
+        if (backgroundColor != other.backgroundColor) return false
```
Comment (jpbarbaud) [inline library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:38]: `When` instead of multiple if ?

Comment (kelianClerc) [reply library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:38]: What about: 
```kotlin
return backgroundColor == other.backgroundColor &&
                contentColor == other.contentColor &&
                disabledBackgroundColor == other.disabledBackgroundColor &&
                disabledContentColor == other.disabledContentColor
```

File: library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt (hunk 1 above)
Comment (kelianClerc) [reply library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:23]: In fact having a if inside a method is not very readable imo, 
is it ok to you if extract it like this : 
```kotlin
@Composable
    override fun backgroundColor(enabled: Boolean): State<Color> {
        val computedColor = if (enabled) backgroundColor else disabledBackgroundColor
        return rememberUpdatedState(computedColor)
    }

```

Comment (jpbarbaud) [reply library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:23]: Ack

File: library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt (hunk 3 above)
Comment (jpbarbaud) [reply library-compose/src/main/java/com/spendesk/grapes/compose/button/GrapesButtonColors.kt:38]: sounds good to me

File: library-compose/src/main/java/com/spendesk/grapes/compose/molecule/AskScreen.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+private fun AskScreen(
+    middlePart: @Composable () -> Unit,
+    approveButton: @Composable () -> Unit,
+    declineButton: @Composable () -> Unit
+) {
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .background(Brush.linearGradient(listOf(GrapesTheme.colors.mainPrimaryDark, GrapesTheme.colors.mainPrimaryNormal)))
+            .padding(16.dp),
```
Comment (jpbarbaud) [inline library-compose/src/main/java/com/spendesk/grapes/compose/molecule/AskScreen.kt]: Not the subject of this PR but this kind of dimens should not be defined in a dimension file ?
like GraoesTheme.spaces.padding.[small|medium|large] ?

Comment (kelianClerc) [reply library-compose/src/main/java/com/spendesk/grapes/compose/molecule/AskScreen.kt]: In fact you are right, we already setup a GrapesTheme.dimens with common spacing dimension used in the project.
I'll update the PR to be use it, good catch thanks 🙏

**Review by Scythe14 — COMMENTED** [review]
Just a small question :)

File: library-compose/src/main/java/com/spendesk/grapes/compose/button/Buttons.kt [hunk 5]
```diff
@@ -1,29 +1,95 @@
...
+    val (buttonColors: ButtonColors, buttonStroke: BorderStroke?) = when (configuration) {
+        GrapesButtonConfiguration.PRIMARY -> Pair(primaryButtonColors(), primaryButtonStroke())
+        GrapesButtonConfiguration.SECONDARY -> Pair(secondaryButtonColors(), secondaryButtonStroke())
+        GrapesButtonConfiguration.TEXT -> Pair(textButtonColors(), textButtonStroke())
+        GrapesButtonConfiguration.ALERT -> Pair(alertButtonColors(), alertButtonStroke())
+        GrapesButtonConfiguration.WARNING -> Pair(warningButtonColors(), warningButtonStroke())
+        GrapesButtonConfiguration.SECONDARY_TEXT -> Pair(secondaryTextButtonColors(), secondaryTextButtonStroke())
+    }
+    when (configuration) {
+        GrapesButtonConfiguration.TEXT, GrapesButtonConfiguration.SECONDARY_TEXT -> {
```
[the remark this answers is not in the document]
Comment (kelianClerc) [reply library-compose/src/main/java/com/spendesk/grapes/compose/button/Buttons.kt:38]: Text is black text as in middle Mais oui
Capture d’écran 2022-10-06 à 11 05 23

Secondary_text is white text as in "Not now"
Capture d’écran 2022-10-06 à 11 05 28

**Review by Scythe14 — APPROVED** [review]
Gooow

**Pull request merged**